src/Document.js

```diff
@@ -1,15 +1,199 @@
-const {HTMLElement, HTMLLoadableElement, Node} = require('./DOM');
+const parse5 = require('parse5');
+
+const DOM = require('./DOM');
+const {Event} = require('./Event');
+const GlobalContext = require('./GlobalContext');
 const symbols = require('./symbols');
+const utils = require('./utils');
+
+const _parseDocument = (s, options, window) => {
+  const ast = parse5.parse(s, {
+    sourceCodeLocationInfo: true,
+  });
+  ast.tagName = 'document';
+  return _parseDocumentAst(ast, options, window, true);
+};
+module.exports._parseDocument = _parseDocument;
+GlobalContext._parseDocument = _parseDocument;
+
+const _parseDocumentAst = (ast, options, window, uppercase) => {
+  const document = GlobalContext._fromAST(ast, window, null, null, uppercase);
+  return initDocument(document, window);
+};
+module.exports._parseDocumentAst = _parseDocumentAst;
+
+/**
+ * Initialize document instance with properties and methods.
+ */
+function initDocument (document, window) {
+  const html = document.childNodes.find(el => el.tagName === 'HTML');
+  const documentElement = html || (document.childNodes.length > 0 ? document.childNodes[0] : null);
+  const head = html ? html.childNodes.find(el => el.tagName === 'HEAD') : null;
+  const body = html ? html.childNodes.find(el => el.tagName === 'BODY') : null;
+
+  document.documentElement = documentElement;
+  document.readyState = 'loading';
+  document.head = head;
+  document.body = body;
+  document.location = window.location;
+  document.createElement = tagName => {
+    tagName = tagName.toUpperCase();
+    const HTMLElementTemplate = window[symbols.htmlTagsSymbol][tagName];
+    const element = HTMLElementTemplate ? new HTMLElementTemplate() : new DOM.HTMLElement(tagName);
+    element.ownerDocument = document;
+    return element;
+  };
+  document.createElementNS = (namespace, tagName) => document.createElement(tagName);
+  document.createDocumentFragment = () => {
+    const documentFragment = new DocumentFragment();
+    documentFragment.ownerDocument = document;
+    return documentFragment;
+  };
+  document.createTextNode = text => new DOM.Text(text);
+  document.createComment = comment => new DOM.Comment(comment);
+  document.createEvent = type => {
+    switch (type) {
+      case 'KeyboardEvent':
+      case 'KeyboardEvents':
+        return new KeyboardEvent();
+      case 'MouseEvent':
+      case 'MouseEvents':
+        return new MouseEvent();
+      case 'Event':
+      case 'Events':
+      case 'HTMLEvents':
+        return new Event();
+      default:
+        throw new Error('invalid createEvent type: ' + type);
+    }
+  };
+  document.createRange
```

Comment:
If the point is a TODO, can we comment that?

Reply:
Oops, accidentally got in there...